File: Threads Posix/exo3.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

struct TypeTableau
{
    int *tab;
    int n;
    int x;
};
typedef struct TypeTableau TypeTableau;


void* initialiser(void* arg) 
{
    TypeTableau *tt = (TypeTableau*)arg;
    tt->tab = malloc(tt->n * sizeof(int));
    srand(time(NULL));
    for (int i = 0; i < tt->n; i++) 
    {
        tt->tab[i] = rand() % 100;
        printf("%d ", tt->tab[i]); 
    }
    printf("\n");
    return NULL;
}


void* chercher(void* arg) 
{
    TypeTableau *tt = (TypeTableau*)arg;
    int *trouve = malloc(sizeof(int));
    *trouve = 0;
    for (int i = 0; i < tt->n; i++) 
    {
        if (tt->tab[i] == tt->x) 
        { 
            *trouve = 1; 
            break; 
        }
    }
    return (void*)trouve;
}

int main() {
    pthread_t t1, t2;
    TypeTableau tt;
    int *resultat;

    printf("Nombre d'éléments : ");
    scanf("%d", &tt.n);

   
    pthread_create(&t1, NULL, initialiser, &tt);

    
    printf("Entrez x à chercher : ");
    scanf("%d", &tt.x);

    
    pthread_join(t1, NULL);

    
    pthread_create(&t2, NULL, chercher, &tt);
    pthread_join(t2, (void**)&resultat);

    printf("Présent ? %d\n", *resultat);

    free(tt.tab);
    free(resultat);
    return 0;
}

File: Threads Posix/exo3.py
import threading
import random
import time
import sys

class TypeTableau:
    def __init__(self):
        self.tab = []
        self.n = 0
        self.x = 0

def initialiser(tt):
    random.seed(time.time())
    tt.tab = [random.randint(0, 99) for _ in range(tt.n)]
    for val in tt.tab:
        print(f"{val} ", end="")
    print()

def chercher(tt, result_container):
    trouve = 0
    for val in tt.tab:
        if val == tt.x:
            trouve = 1
            break
    result_container[0] = trouve

def main():
    tt = TypeTableau()

    tt.n = int(input("Nombre d'éléments : "))

    t1 = threading.Thread(target=initialiser, args=(tt,))
    t1.start()

    tt.x = int(input("Entrez x à chercher : "))

    t1.join()

    resultat = [0]
    t2 = threading.Thread(target=chercher, args=(tt, resultat))
    t2.start()
    t2.join()

    print(f"Présent ? {resultat[0]}")

if __name__ == "__main__":
    main()